Dataops 1083 pUC19 lambda plots

sarek928 opened the pull request:
I have translated R scripts to python. Please be thorough since copilot has been very much involved.

sarek928 committed:
Add scripts for plotting pUC19 and lambda for methylseq data

sarek928 committed:
README updated with reference to methylation scripts

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+
+import os
+import pandas as pd
+import matplotlib.pyplot as plt
```
matplotlib is not automatically available in the NGI environment on miarka so we need to remember to add instructions in confluence to run `module load python` before running these scripts.

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+parser = argparse.ArgumentParser(
+    description="This script finds methylation data in the multiqc results directory created by nf-core methylseq and plots the methylation rate in CpG sites for lambda."
+)
+parser.add_argument("dir", type=str, help="path to analysis directory")
```
I think the help text could be more informative. maybe just add "path to lambda control analysis directory" or something similar

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+
+# Find the paths to the two files from the provided working directory
+dir = args.dir
+file1_path = os.path.join(
```
This works, but I think it's easier to follow with more specific variable names, e.g. file1_path --> alignment_stats

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+file1_path = os.path.join(
+    dir, "results", "multiqc", "bismark", "multiqc_data", "mqc_bismark_alignment_1.txt"
+)
+file2_path = os.path.join(
```
same here as above, maybe methylation_stats ?

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+
+# Check if the files exist
+if not os.path.exists(file1_path) or not os.path.exists(file2_path):
+    print("One or both of the files do not exist. Please check the paths.")
```
Kind of hard to know which paths to check unless you go in and read the code. Maybe print the paths for file1 and file2 in the error message?

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+args = parser.parse_args()
+
+# Find the paths to the two files from the provided working directory
+dir = args.dir
```
dir is a built in function in python, so it would be good to rename this variable.

sarek928 replied:
changed to analysis_dir

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+data2_selected.loc[:, "Sample"] = data2_selected["Sample"].str.replace("_1_val_1", "")
+
+# Merge the two DataFrames based on the common column (Sample)
+merged_data = pd.merge(data1_selected, data2_selected, on="Sample", how="left")
```
This can be written more concise. Also, it's safer to get the data by column name than column number in case the column order is updated in the input file: 

```diff
- # Read the first file with header
- data1 = pd.read_csv(file1_path, delimiter="\t")

- # Read the second file with header
- data2 = pd.read_csv(file2_path, delimiter="\t")

- # Select columns by index (assuming 0-based index)
- # Column 0: Sample (from both files)
- # Column 1: Reads (from file 1)
- # Column 6: %CpG (from file 2)
- data1_selected = data1.iloc[:, [0, 1]].copy()
- data2_selected = data2.iloc[:, [0, 6]].copy()

- # Rename columns to desired names
- data1_selected.columns = ["Sample", "Reads"]
- data2_selected.columns = ["Sample", "%CpG"]

- # Trim the suffix '_1_val_1' from the 'Sample' column using .loc
- data1_selected.loc[:, "Sample"] = data1_selected["Sample"].str.replace("_1_val_1", "")
- data2_selected.loc[:, "Sample"] = data2_selected["Sample"].str.replace("_1_val_1", "")

- # Merge the two DataFrames based on the common column (Sample)
- merged_data = pd.merge(data1_selected, data2_selected, on="Sample", how="left")
+ data1 = pd.read_csv(file1_path, delimiter="\t", usecols=["Sample", "Aligned Uniquely"])
+ data2 = pd.read_csv(file2_path, delimiter="\t", usecols=["Sample", "percent_cpg_meth"])
+ merged_data = pd.merge(data1, data2, on="Sample", how="left")
+ merged_data["Sample"] = merged_data["Sample"].str.replace("_1_val_1", "")
+ merged_data.columns = ["Sample", "Reads", "%mCpG"]
```

sarek928 replied:
I agree, so much easier to read.

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+
+
+# Define colors based on %CpG values
+def get_color(value):
```
Move this function to just after the imports in the beginning of the code

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+        return "red"
+
+
+colors = merged_data["%CpG"].apply(get_color)
```
```suggestion
colors = merged_data["%mCpG"].apply(get_color)
```

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+plt.xticks(rotation=90)
+
+# Add %CpG values on top of the bars if they are over 0.7
+for bar, value in zip(bars, merged_data["%CpG"]):
```
```suggestion
for bar, value in zip(bars, merged_data["%mCpG"]):
```

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+        )
+
+# Add legend
+from matplotlib.lines import Line2D
```
Move to the other import in the begging of the code

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+    Line2D([0], [0], color="red", lw=4, label="Failed"),
+]
+
+plt.legend(handles=legend_elements, loc="upper right")
```
This works but sometimes the legend box might overlap with the data.
This can be solved by putting it outside the plot are with something like:

```suggestion
plt.legend(handles=legend_elements, loc='center left', bbox_to_anchor=(1.02, 0.5), title="%mCpG QC")
```

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+plt.legend(handles=legend_elements, loc="upper right")
+
+# Save plot to a file in the provided working directory
+plt.tight_layout()
```
If you want to do the changes above to the legend box, also include this:

```suggestion
plt.tight_layout(rect=[0,0,1,1])
```

kjellinjonas commented on plot_lambda.py:
```diff
@@ -0,0 +1,114 @@
+
+# Save plot to a file in the provided working directory
+plt.tight_layout()
+plt.savefig(os.path.join(dir, "lambda.png"))
```
if you want to include the extra parameter mentioned above, you can have a more specific file name here and in the print statement below:

```suggestion
plt.savefig(os.path.join(dir, f"{prefix}_lambda.png}")
```

kjellinjonas reviewed (changes requested):
Nice work! The scripts are doing what they are supposed to do and I think the plot layouts are working well. I would maybe prefer to have the same colors for pass and fail for both plots, but it's not important.

I made a lot of comments on code structure, naming of variables etc on the lambda script but they apply to the pUC19 script as well. Let me know if you want to discuss any of the proposed changes!

sarek928 committed:
Update plot_lambda.py

Updated script based on suggestions.

sarek928 committed:
Update plot_pUC19.py

Updated script based on suggestions.

sarek928 commented:
@kjellinjonas  OK, so I have updated the files now. I have no idea if this was the correct way of doing things. I just pressed edit and updated the files in the GUI. But feel free to check it out and merge if ok.
